File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.cpp
// Last updated: 17/07/2025, 20:52:01
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int helper(TreeNode* root, int& res){
        
        if(root == NULL){
            return 0;
        }
        int lh = helper(root->left, res) ;
        int rh = helper(root->right, res);
        res = max(lh+rh,res);
        // cout<< max(lh, rh)<<" " << root->val<<endl;
        return max(lh, rh) + 1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int res  = 0;
        helper(root, res);
        return res;

    }
};

File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.py
# Last updated: 01/10/2025, 00:12:49
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_diameter = 0
        def depth(node):
            if not node:
                return 0
            left_depth = depth(node.left)
            right_depth = depth(node.right)

            self.max_diameter = max(self.max_diameter, left_depth + right_depth)

            return 1 + max(left_depth, right_depth)
        depth(root)
        return self.max_diameter
        